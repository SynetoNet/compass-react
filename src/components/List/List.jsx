import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import ListItem from "./components/ListItem";
import ListActions from "./components/ListActions";

import "./List.scss";

const List = ({
  title,
  tabbed,
  compact,
  actions,
  items,
  children,
  className = "",
  ...props
}) => {
  const classes = (
    classNames({
      compact,
      actions
    }) + ` list-wrapper ${className}`
  ).trim();

  if (!items && !children) {
    throw `List component require either the "items" prop, or the "children" prop`;
  }

  const content = children || getContent(items);
  const ListElement = tabbed ? "dl" : "ul";

  return (
    <div className={classes}>
      {renderTitle()}
      {renderActions()}
      <ListElement className="list" {...props}>
        {content}
      </ListElement>
    </div>
  );

  function getContent(items) {
    return items.map(({ value, label }, index) => {
      const key = typeof label === "string" ? label : index;
      const _actions = !compact ? actions : undefined;
      return (
        <ListItem key={key} label={label} tabbed={tabbed} actions={_actions}>
          {value}
        </ListItem>
      );
    });
  }

  function renderTitle() {
    return title ? (
      <header className="list_header">
        <h4 className="list_title">{title}</h4>
      </header>
    ) : null;
  }

  function renderActions() {
    if (!actions || !compact) {
      return null;
    }

    return <ListActions actions={actions} single />;
  }
};

List.Item = ListItem;

List.propTypes = {
  title: PropTypes.string,
  compact: PropTypes.bool,
  tabbed: PropTypes.bool,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.node,
      value: PropTypes.node
    })
  ),
  children: PropTypes.oneOfType([
    PropTypes.shape({
      type: ListItem
    }),
    PropTypes.arrayOf(
      PropTypes.shape({
        type: ListItem
      })
    )
  ]),
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      onSelect: PropTypes.func
    })
  )
};

List.defaultProps = {
  compact: false,
  tabbed: false
};

export default List;
