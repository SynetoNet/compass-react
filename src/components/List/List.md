#### **Basic List**

Without title:

```jsx
<List
  items={[
    { label: "Phone", value: "(0756) 789 123" },
    { label: "Email Address", value: "jonsnow@winterfell.north" }
  ]}
/>
```

With title:

```jsx
<List
  title="List title"
  items={[
    { label: "Phone", value: "(0756) 789 123" },
    { label: "Email Address", value: "jonsnow@winterfell.north" }
  ]}
/>
```

#### **Tabbed List**

```jsx
<List
  title="List title"
  tabbed
  items={[
    { label: "Phone", value: "(0756) 789 123" },
    { label: "Email Address", value: "jonsnow@winterfell.north" }
  ]}
/>
```

#### **Compact List**

```jsx
<List
  title="List title"
  compact
  items={[
    { label: "Phone", value: "(0756) 789 123" },
    { label: "Email Address", value: "jonsnow@winterfell.north" }
  ]}
/>
```

#### **List item actions**

```jsx
<List
  title="List title"
  actions={[
    {
      label: "Edit",
      onSelect: value => console.log("Edit", value)
    },
    {
      label: "Delete",
      onSelect: value => console.log("Delete", value)
    }
  ]}
  items={[
    { label: "Phone", value: "(0756) 789 123" },
    { label: "Email Address", value: "jonsnow@winterfell.north" }
  ]}
/>
```

#### **Actions in compact list**

Without title:

```jsx
<List
  compact
  actions={[
    {
      label: "Edit",
      onSelect: value => console.log("Edit", value)
    },
    {
      label: "Delete",
      onSelect: value => console.log("Delete", value)
    }
  ]}
  items={[
    { label: "Phone", value: "(0756) 789 123" },
    { label: "Email Address", value: "jonsnow@winterfell.north" }
  ]}
/>
```

With title:

```jsx
<List
  title="List title"
  compact
  actions={[
    {
      label: "Edit",
      onSelect: value => console.log("Edit", value)
    },
    {
      label: "Delete",
      onSelect: value => console.log("Delete", value)
    }
  ]}
  items={[
    { label: "Phone", value: "(0756) 789 123" },
    { label: "Email Address", value: "jonsnow@winterfell.north" }
  ]}
/>
```

#### **Extended usage**

```jsx
<List title="List title">
  <List.Item label="Phone">(0756) 789 123</List.Item>
  <List.Item label="Email Address">jonsnow@winterfell.north</List.Item>
</List>
```
