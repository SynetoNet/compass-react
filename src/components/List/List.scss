@import "../../base/colors";
@import "../../base/variables";

.list-wrapper {
  position: relative;
}

ul.list {
  margin: 0;
  list-style-type: none;
  padding: 0;
}

dl.list {
  display: grid;
  grid-template-columns: minmax(2rem, max-content) 1fr;
  margin: 0;
}

.list-header {
  display: flex;
  align-items: center;
  padding-bottom: 0.75rem;

  .list-wrapper:not(.compact) & {
    border-bottom: 1px $neutral-6 solid;
  }
}

.list-title {
  flex: 1;
  margin: 0;
}

.list-item {
  margin: 0;
  padding: 0.15rem 0;

  .list-wrapper:not(.compact) & {
    padding-top: 0.75rem;
    padding-bottom: 0.75rem;
  }

  .list-wrapper.compact.actions & {
    max-width: calc(100% - 30px);
  }
}

.list-wrapper:not(.compact) .list-item:not(:last-of-type) {
  border-bottom: 1px $neutral-6 solid;
}

.list-value {
  padding-left: 1em;
  color: $neutral-40;
}

li.list-item {
  display: flex;
  align-items: center;

  .list-value {
    flex: 1;
  }
}

.list-label {
  font-weight: bold;
}

dd.list-value {
  display: flex;
  justify-content: space-between;
}

.list-actions {
  visibility: hidden;
  padding-left: 1em;

  .list-wrapper.compact.actions:hover & {
    visibility: visible;
  }

  .list-wrapper.actions .list-item:hover & {
    visibility: visible;
  }
}

.list-actions--single {
  position: absolute;
  top: 0.25em;
  right: 0;
}
