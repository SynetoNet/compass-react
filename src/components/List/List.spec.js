import React from "react";
import List from "./List";
import renderer from "react-test-renderer";

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error) {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    // You can also log the error to an error reporting service
    console.log(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      // You can render any custom fallback UI
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children; 
  }
}

test("should throw if no items or children are passed", () => {
  expect(() => renderer.create(<List />)).toThrow();
});

test("rendering same output using items prop of children", () => {
  const withItems = renderer.create(
    <List
      items={[
        { label: "label 1", value: "value 1" },
        { label: "label 2", value: "value 2" }
      ]}
    />
  );

  const withChildren = renderer.create(
    <List>
      <List.Item label="label 1">value 1</List.Item>
      <List.Item label="label 2">value 2</List.Item>
    </List>
  );

  expect(withItems.toJSON()).toEqual(withChildren.toJSON());
});

test("renders List with title", () => {
  const component = renderer.create(
    <List title="list title">
      <List.Item label="label 1">value 1</List.Item>
      <List.Item label="label 2">value 2</List.Item>
    </List>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test("renders tabbled List", () => {
  const component = renderer.create(
    <List tabbed>
      <List.Item label="label 1">value 1</List.Item>
      <List.Item label="label 2">value 2</List.Item>
    </List>
  );
  expect(component.toJSON()).toMatchSnapshot();
});
