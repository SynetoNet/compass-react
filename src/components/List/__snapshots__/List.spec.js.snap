// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders List with title 1`] = `
<div
  className="list-wrapper"
>
  <header
    className="list-header"
  >
    <h4
      className="list-title"
    >
      list title
    </h4>
  </header>
  <ul
    className="list"
    compact={false}
  >
    <li
      className="list-item"
    >
      <div
        className="list-label"
      >
        label 1
      </div>
      <div
        className="list-value"
      >
        value 1
      </div>
    </li>
    <li
      className="list-item"
    >
      <div
        className="list-label"
      >
        label 2
      </div>
      <div
        className="list-value"
      >
        value 2
      </div>
    </li>
  </ul>
</div>
`;

exports[`renders tabbled List 1`] = `
<div
  className="list-wrapper"
>
  <dl
    className="list"
    compact={false}
  >
    <li
      className="list-item"
    >
      <div
        className="list-label"
      >
        label 1
      </div>
      <div
        className="list-value"
      >
        value 1
      </div>
    </li>
    <li
      className="list-item"
    >
      <div
        className="list-label"
      >
        label 2
      </div>
      <div
        className="list-value"
      >
        value 2
      </div>
    </li>
  </dl>
</div>
`;
